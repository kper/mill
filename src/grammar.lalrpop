use std::str::FromStr;
use crate::ast::*;
use lalrpop_util::ParseError;
use anyhow::Error;

grammar;

extern {
    type Error = Error;
}

Comma<T>: Vec<T> = { 
    <v:(<T> ",")*> <e:T?> => match e {
        None => v,
        Some(e) => {
            let mut v = v;
            v.push(e);
            v
        }
    }
};

SemiComma<T>: Vec<T> = { 
    <v:(<T> ";")*> => v
};

Vec<T>: Vec<T> = {
    <v:(<T>)*> => v
};

pub Pars  = Comma<Par>;

Par : Identifier = {
    <id:Id> ":" <dat:DataType> => {
        id.update_ty(dat)
    }
};
pub Program : Program = <decls:Vec<Decl>> => {
    Program {
        functions: decls.clone().into_iter()
                        .filter_map(|w| match w { Decl::Func(f) => Some(f), _ => None })
                        .collect(),
        structs: decls.clone().into_iter()
                        .filter_map(|w| match w { Decl::Struct(f) => Some(f), _ => None })
                        .collect()
    }
};

pub Decl: Decl = {
    <f:Funcdef> => Decl::Func(f),
    <s:Struct> => Decl::Struct(s)
};

pub Funcdef : Func = {
	"fn" <id:Id> "(" <pars:Pars> ")" "{" <statements:Statements> "}" =>? Func::new(id, pars, statements).map_err(|w| ParseError::User {
        error: w
    })
};

pub Struct : Struct = {
    "struct" <id:Id> "{" <fields:Comma<Field>> "}" =>? Struct::new(id, fields).map_err(|w| ParseError::User {
        error: w
    })
};

Field: Field = {
    <id:Id> ":" <dat:DataType> => Field::new(id, dat)
};

DataType: DataType = {
    "int" => DataType::Int,
    <id:Id> => DataType::Struct(Box::new(id))
}

pub Statements = SemiComma<Statement>;

pub Statement: Box<Statement> = {
	"return" <e:Expr> => Box::new(Statement::Ret(e)),
    "let" <id:Id> ":" <dat:DataType> "=" <e:Expr> => {
        Box::new(Statement::Assign(id.update_ty(dat), e))
    },
    <id:Id> ":" <cond:Conditional> => Box::new(Statement::Conditional(Some(id), cond)),
    <cond:Conditional> => Box::new(Statement::Conditional(None, cond)),
    <cond:Conditional> <id:Id> ":" => Box::new(Statement::Conditional(Some(id), cond)),
    <id:Id> "=" <e:Expr> => Box::new(Statement::ReAssign(id, e)),
    <id:Id> "." <field:Id> "=" <e:Expr> => 
        Box::new(Statement::ReAssign(id.update_field_access(Some(field)), e)),
	"let" <id:Id> "=" "new" <st:Id> =>
        Box::new(Statement::Allocate(id.update_ty(DataType::Struct(Box::new(st.clone()))), st))
};

pub Conditional: Vec<Box<Guard>> = {
    "match" <guards:SemiComma<Guard>> "end" => guards
};

pub Guard: Box<Guard> = {
    <e:Expr> "->" <stats:Statements> <cont:Continuation> => Box::new(Guard {
        guard: Some(e),
        statements: stats,
        continuation: cont,
    }),
    "_" "->" <stats:Statements> <cont:Continuation> => Box::new(Guard {
        guard: None,
        statements: stats,
        continuation: cont,
    })
};

Continuation: Continuation = {
    "continue" => Continuation::Continue,
    "break" => Continuation::Break,
};

pub Expr: Box<Expr> = {
    <n1:Term> <op:BinOp> <n2:Term> => Box::new(Expr::Dual(op, n1, n2)),
    "(" <n1:Term> <op:BinOp> <n2:Term> ")" => Box::new(Expr::Dual(Opcode::Geq, n1, n2)),
};

Exprs: Vec<Box<Expr>> = Comma<Expr>;

BinOp: Opcode = {
    "==" => Opcode::Cmp
};

pub Term: Box<Term> = {
	Num => Box::new(Term::Num(<>)),
	Id => Box::new(Term::Id(<>)),
    <obj:Id> "." <field:Id> => 
        Box::new(Term::Object(obj.update_field_access(Some(field.clone())), field)),
	"(" <n:Term> ")" => n,
	<id:Id> "(" <exprs:Exprs> ")" => Box::new(Term::Call(id, exprs)),
};

pub Id: Identifier = <l:@L> <s:r"[a-zA-Z][a-zA-Z0-9]*"> <r:@R> => Identifier::new(s.to_string(), l, r, None);
Num: i64= <s:r"[0-9]+"> => i64::from_str(s).unwrap();
