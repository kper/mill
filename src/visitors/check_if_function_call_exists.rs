use crate::visitors::*;
use crate::ast::*;

use crate::symbol_table::SymbolTable;

use anyhow::bail;

#[derive(Debug, Default)]
pub struct CheckIfFunctionCallExistsVisitor {
    symbol_table: SymbolTable
}

impl Visitor for CheckIfFunctionCallExistsVisitor {
    fn visit_program<'ctx>(&mut self, program: &'ctx Program) -> Result<()> {
        //println!("Visiting program");
        Ok(())
    }

    fn visit_func<'ctx>(&mut self, func: &'ctx Func) -> Result<()> {
        //println!("Visiting func");
        Ok(())
    }

    fn visit_statement<'ctx>(&mut self, stmt: &'ctx Statement) -> Result<()> {
        //println!("Visiting statement");
        Ok(())
    }
    
    //fn visit_guard(&mut self, label: &Option<IdTy>, guard: &mut Guard) -> Result<()>;

    fn visit_guard<'ctx>(&mut self, guard: &'ctx Guard) -> Result<()> {
        //println!("Visiting guard");
        Ok(())
    }

    fn visit_expr<'ctx>(&mut self, expr: &'ctx Expr) -> Result<()> {
        //println!("Visiting expr");
        Ok(())
    }

    fn visit_term<'ctx>(&mut self, term: &'ctx Term) -> Result<()> {
        let symbol_table = &self.symbol_table;
        match term {
            Term::Call(id, exprs) => {
                if !symbol_table.lookup_symbol(&id.get_name()) {
                    bail!("Function {} is not defined", id);
                }
            }
            _ => {}
        }

        Ok(())
    }

    fn visit_struct<'ctx>(&mut self, stru: &'ctx Struct) -> Result<()> {
        //println!("Visiting struct");
        Ok(())
    }
}