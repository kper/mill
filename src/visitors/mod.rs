use crate::ast::*;
use crate::symbol_table::SymbolTable;
use anyhow::Result;
use inkwell::values::{BasicValueEnum, FunctionValue};
use std::borrow::Cow;

pub trait CheckIfFunctionCallExistsVisitor {
    fn lookup(functions: &SymbolTable, name: &IdTy) -> bool {
        functions.lookup_symbol(name)
    }

    fn visit(&self, functions: &SymbolTable) -> Result<bool>;
}

pub trait CodegenVisitor<'ctx> {
    fn visit_program(&mut self, program: &'ctx mut Program) -> Result<()>;
    fn visit_func(&mut self, func: &'ctx mut Func) -> Result<()>;
    fn visit_statement(&mut self, stmt: &Statement, function: &IdTy) -> Result<()>;
    fn visit_guard(&mut self, label: &Option<IdTy>, guard: &Guard, function: &IdTy) -> Result<()>;
    fn visit_expr(&mut self, expr: &Expr) -> Option<Cow<BasicValueEnum<'ctx>>>;
    fn visit_term(&mut self, term: &Term) -> Option<Cow<BasicValueEnum<'ctx>>>;
}
