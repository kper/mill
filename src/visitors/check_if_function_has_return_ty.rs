use crate::visitors::*;

use anyhow::{bail};

#[derive(Debug, Default)]
pub struct CheckIfFunctionHasReturnTyVisitor;

impl Visitor for CheckIfFunctionHasReturnTyVisitor {
    fn get_name(&self) -> String {
        "CheckIfFunctionHasReturnTyVisitor".to_string()
    }

    fn visit_program<'a>(&mut self, _ctx: Ctx<'a>, _program: &Program) -> Result<()> {
        Ok(())
    }

    fn visit_func<'a>(&mut self, _ctx: Ctx<'a>, _func: &Func) -> Result<()> {
        Ok(())
    }

    fn visit_statement<'a>(&mut self, ctx: Ctx<'a>, stmt: &Statement) -> Result<()> {
        if let Some(function) = ctx.function {
            match stmt {
                Statement::RetVoid => {
                    if function.ret_ty.is_some() {
                        bail!("Declared function's return type is not void. But the return statement does not return anything.");
                    }
                }
                Statement::Ret(_exprs) => {
                    // TODO add type checks if the ret ty is correct.
                    if function.ret_ty.is_none() {
                        bail!("Declared function's return type is void. But a variable is returned.");
                    }
                }
                _ => {}
            }
        }

        Ok(())
    }

    fn visit_expr<'a>(&mut self, _ctx: Ctx<'a>, _expr: &Expr) -> Result<()> {
        Ok(())
    }

    fn visit_term<'a>(&mut self, _ctx: Ctx<'a>, _term: &'a Term) -> Result<()> {
        Ok(())
    }

    fn visit_struct<'a>(&mut self, _ctx: Ctx<'a>, _stru: &'a Struct) -> Result<()> {
        Ok(())
    }
}
