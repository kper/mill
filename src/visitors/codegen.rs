use anyhow::Result;

use crate::Visitor;
use crate::visitors::CodegenVisitorTrait;
use crate::codegen::Codegen;
use crate::ast::*;
use log::{debug, warn};

use llvm_sys::bit_writer::*;
use llvm_sys::core::*;
use std::ffi::CString;
use std::ptr;
use crate::c_str;

use crate::symbol_table::*;

pub struct CodegenVisitor {
}

impl CodegenVisitor {
    pub fn new() -> Self {
        Self {
        }
    }
}

impl CodegenVisitorTrait for CodegenVisitor {
    fn get_name(&self) -> String {
        "CodegenVisitor".to_string()
    }

    fn visit_program(&mut self, _program: &Program, codegen: &mut Codegen) -> Result<()> {
        debug!("{}: running visit_program", self.get_name());
        Ok(())
    }

    fn visit_func(&mut self, func: &Func, codegen: &mut Codegen) -> Result<()> {
        debug!("{}: running visit_func", self.get_name());

        let context = codegen.context.clone();
        let module = codegen.module.clone();
        let builder = codegen.builder.clone();

        unsafe {
            let void_type = LLVMVoidTypeInContext(context);
            let i8_type = LLVMIntTypeInContext(context, 8);
            let _i8_pointer_type = LLVMPointerType(i8_type, 0);

            let func_type = LLVMFunctionType(void_type, ptr::null_mut(), 0, 0);
            let func_llvm = LLVMAddFunction(module, c_str!(func.id.get_name()), func_type);
            let block = LLVMAppendBasicBlockInContext(context, func_llvm, c_str!(func.id.get_name()));
            LLVMPositionBuilderAtEnd(builder, block);
            
            codegen.block_table.insert(func.id.get_name(), block)?;
            codegen.function_table.insert(&func.id.get_name(), func_llvm)?;

            //let hello_world_str = LLVMBuildGlobalStringPtr(builder, c_str!("hello, world."), c_str!(""));

            //LLVMBuildRetVoid(builder);
        }

        //let context = codegen.get_context();

        //let builder = codegen.get_mut_builder();
        //let module = codegen.get_mut_module();
        //let context = codegen.get_context();
        //let ftable = codegen.get_mut_function_table();
        //let symbol_table = codegen.get_mut_symtable();

        /*
        let context = codegen.context;
        let i64_type = context.i64_type();
        let func_types = vec![i64_type.into(); func.pars.len()]; 
        let fn_type = i64_type.fn_type(&func_types, false);

        let function = codegen.get_mut_module().add_function(&func.id.get_name(), fn_type, None);

        codegen.get_mut_function_table().insert(&func.id.get_name(), function)?;

        // Basic block

        let basic = context.append_basic_block(function, func.id.get_name());

        builder.position_at_end(basic);

        for (i, param) in func.pars.iter().enumerate() {
            let value = function.get_nth_param(i as u32).unwrap();
            let ptr = builder.build_alloca(i64_type, param.get_name());

            let _instr = builder.build_store(ptr, value);

            symbol_table.insert(
                param.get_name(),
                (param.clone(), BasicValueEnum::PointerValue(ptr)),
            )?;
            debug!("Allocating functions parameter {}", param);
        }
        */

        Ok(())
    }

    fn visit_statement(&mut self, stmt: &Statement, codegen: &mut Codegen) -> Result<()> {
        debug!("{}: Visiting statement {:#?}", self.get_name(), stmt);

        //TODO add return void
        let context = codegen.context.clone();
        let module = codegen.module.clone();
        let builder = codegen.builder.clone();


        match stmt {
            Statement::Ret(_expr) => {
                //unimplemented!()

                unsafe {
                    // this is wrong
                    LLVMBuildRetVoid(builder);
                }
            }
            Statement::Assign(id, _expr) => {
                let sym_expr = codegen.expr_table.get_last(); 

                if let Some(value) = sym_expr {
                    debug!("Building bit cast for {}", id.get_name());

                    unsafe {
                        let i64_ty = LLVMIntTypeInContext(codegen.context, 64);
                        let ty = BasicValueType::Int(i64_ty);

                        let _ = value.store(context, builder, id)?;

                        codegen.symbol_table.insert(id.get_name(), (id.clone(), BasicValue {
                            ty: ty.clone(),
                            value: value.value,
                        }))?;
                    }
                }
                else {
                    warn!("No last sym for the assignment statement");
                }
            }
            _ => {

            }
        }

        Ok(())
    }
    
    fn visit_guard(&mut self, _guard: &Guard, codegen: &mut Codegen) -> Result<()> {
        Ok(())
    }

    fn visit_expr(&mut self, expr: &Expr, codegen: &mut Codegen) -> Result<()> {
        debug!("{}: Visiting expr {:#?}", self.get_name(), expr);

        match expr {
            Expr::Single(_term) => {
                // do nothing
            }
            Expr::Num(num) => {
                unsafe {
                    let i64_ty = LLVMIntTypeInContext(codegen.context, 64);
                    let value = LLVMConstInt(i64_ty, *num as u64, 0);
                    codegen.expr_table.push(value, BasicValueType::Int(i64_ty))?;
                }
            }
            _ => {
                unimplemented!()
            }
        }

        Ok(())
    }

    fn visit_term(&mut self, term: &Term, codegen: &mut Codegen) -> Result<()> {
        debug!("{}: Visiting term {:#?}", self.get_name(), term);

        match term {
            Term::Num(num) => {
                unsafe {
                    let i64_ty = LLVMIntTypeInContext(codegen.context, 64);
                    let value = LLVMConstInt(i64_ty, *num as u64, 0);
                    codegen.expr_table.push(value, BasicValueType::Int(i64_ty))?;
                }
            }
            _ => {
                unimplemented!();
            }
        }
        
        Ok(())
    }

    fn visit_struct(&mut self, _stru: &Struct, codegen: &mut Codegen) -> Result<()> {
        Ok(())
    }

    fn write_bitcode(&self, name: &str) -> Result<bool> {
        //self.codegen.write_bitcode(name)?;
        Ok(true)
    }

    fn get_ir(&self) -> Result<Option<String>> {
        //Ok(Some(self.codegen.get_ir()))
        unimplemented!()
    }
}