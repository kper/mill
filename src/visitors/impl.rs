use crate::visitors::*;
use crate::ast::*;
use either::Either;

pub struct PrintVisitor;
pub struct NormalTraversal;

impl<'ctx> Visitor<'ctx> for PrintVisitor{
    fn visit_program(&mut self, program: &'ctx mut Program) -> Result<()> {
        println!("Visiting program");
        Ok(())
    }

    fn visit_func(&mut self, func: &'ctx mut Func) -> Result<()> {
        println!("Visiting func");
        Ok(())
    }

    fn visit_statement(&mut self, stmt: &mut Statement) -> Result<()> {
        println!("Visiting statement");
        Ok(())
    }
    
    //fn visit_guard(&mut self, label: &Option<IdTy>, guard: &mut Guard) -> Result<()>;

    fn visit_guard(&mut self, guard: &mut Guard) -> Result<()> {
        println!("Visiting guard");
        Ok(())
    }

    fn visit_expr(&mut self, expr: &mut Expr) -> Result<()> {
        println!("Visiting expr");
        Ok(())
    }

    fn visit_term(&mut self, term: &mut Term) -> Result<()> {
        println!("Visiting term");
        Ok(())
    }

    fn visit_struct(&mut self, stru: &mut Struct) -> Result<()> {
        println!("Visiting struct");
        Ok(())
    }
}

impl Traversal for NormalTraversal {
    fn traverse<'ctx>(&mut self, visitor: &mut Box<dyn Visitor<'ctx>>, program: &'ctx mut Program) -> Result<()> {
           visitor.visit_program(program)?;

            for struc in program.structs.iter_mut() {
                visitor.visit_struct(struc)?;
            }
            
            for function in program.functions.iter_mut() {
                visitor.visit_func(function)?;

                for statement in function.statements.iter_mut() {

                    let expr_or_guard = statement.get_mut_inner();

                    if let Some(expr_or_guard) = expr_or_guard {
                        match expr_or_guard {
                            Either::Left(expr) => {
                                visitor.visit_expr(expr)?;
                            }
                            Either::Right(guards) => {
                                for guard in guards {
                                    visitor.visit_guard(guard)?;
                                } 
                            }
                        }
                    }

                    visitor.visit_statement(statement.as_mut())?;
                    
                }
            }

        Ok(())
    }
}