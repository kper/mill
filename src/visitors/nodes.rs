use crate::visitors::*; 
use crate::ast::*;

impl<'ctx> AbstractNode<'ctx> for Program {
    fn accept(&'ctx mut self, visitor: &mut impl Visitor<'ctx>) {
        visitor.visit_program(self);
    }
}

impl<'ctx> AbstractNode<'ctx> for Func {
    fn accept(&'ctx mut self, visitor: &mut impl Visitor<'ctx>) {
        visitor.visit_func(self);
    }
}

impl<'ctx> AbstractNode<'ctx> for Statement {
    fn accept(&'ctx mut self, visitor: &mut impl Visitor<'ctx>) {
        visitor.visit_statement(self);
    }
}

impl<'ctx> AbstractNode<'ctx> for Guard {
    fn accept(&'ctx mut self, visitor: &mut impl Visitor<'ctx>) {
        visitor.visit_guard(self);
    }
}

impl<'ctx> AbstractNode<'ctx> for Expr {
    fn accept(&'ctx mut self, visitor: &mut impl Visitor<'ctx>) {
        visitor.visit_expr(self);
    }
}

impl<'ctx> AbstractNode<'ctx> for Term {
    fn accept(&'ctx mut self, visitor: &mut impl Visitor<'ctx>) {
        visitor.visit_term(self);
    }
}

impl<'ctx> AbstractNode<'ctx> for Struct {
    fn accept(&'ctx mut self, visitor: &mut impl Visitor<'ctx>) {
        visitor.visit_struct(self);
    }
}