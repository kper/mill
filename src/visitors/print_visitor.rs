use crate::visitors::*;

pub struct PrintVisitor;

impl Visitor for PrintVisitor {
    fn get_name(&self) -> String {
        "PrintVisitor".to_string()
    }

    fn visit_program(&mut self, _program: &Program) -> Result<()> {
        println!("Visiting program");
        Ok(())
    }

    fn visit_func(&mut self, _func: &Func) -> Result<()> {
        println!("Visiting func");
        Ok(())
    }

    fn visit_statement(&mut self, _stmt: &Statement) -> Result<()> {
        println!("Visiting statement");
        Ok(())
    }

    fn visit_expr(&mut self, _expr: &Expr) -> Result<()> {
        println!("Visiting expr");
        Ok(())
    }

    fn visit_term(&mut self, _term: &Term) -> Result<()> {
        println!("Visiting term");
        Ok(())
    }

    fn visit_struct(&mut self, _stru: &Struct) -> Result<()> {
        println!("Visiting struct");
        Ok(())
    }
}
