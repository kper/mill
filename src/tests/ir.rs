use crate::grammar;
use insta::assert_snapshot;
use crate::visitors::CheckIfFunctionCallExistsVisitor;

fn init() {
    let _ = env_logger::builder().is_test(true).try_init();
}

macro_rules! compile {
    ($input:expr) => {
        init();

        let input = $input;

        let mut program = grammar::ProgramParser::new().parse(&input).unwrap();
        let functions = program.get_function_names().unwrap();

        program.visit(&functions).unwrap();

        let ir = program.codegen_to_ir().unwrap();

        assert_snapshot!(ir);
    };
}

#[test]
fn test_return() {
    compile!("fn main() { return 1; }");
}

#[test]
fn test_assignment() {
    compile!("fn main() { let a : int = 1; let b : int = 2; }");
}

#[test]
fn test_reassignment() {
    compile!("fn main() { let a : int = 1; a  = 2; }");
}

#[test]
fn test_addition() {
    compile!("fn main() { return 1 + 2; }");
}

#[test]
fn test_multiple_addition() {
    compile!("fn main() { return 1 + 2 + 3; }");
}

#[test]
fn test_multiple_addition_with_vars() {
    compile!("fn main() { let a : int = 1; let b : int = 2; let c : int = 3; return a + b + c; }");
}

#[test]
fn test_call_when_names_in_order() {
    compile!("fn f(a: int) { return a; } fn main() { let a : int = 1; return f(a); }");
}

#[test]
fn test_call_when_names_not_in_order() {
    compile!("fn main() { let a : int = 1; return f(a); } fn f(a:int) { return a; }");
}

#[test]
fn test_one_guard_with_expr() {
    compile!("fn main() { match 1 -> return 2; break; end; }");
}

#[test]
fn test_mixed_guards() {
    compile!("fn main() { match 1 -> let a : int = 2; break; _ -> return 3; break; end; }");
}

#[test]
fn test_struct() {
    compile!("struct test { a: int }");
}

#[test]
fn test_alloc_struct() {
    compile!("struct test { a: int, b: int } fn main() { let a = new test; }");
}