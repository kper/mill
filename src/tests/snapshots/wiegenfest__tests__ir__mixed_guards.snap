---
source: src/tests/ir.rs
expression: ir

---
; ModuleID = 'main'
source_filename = "main"

define i64 @main() {
main:
  br label %"0"

"0":                                              ; preds = %main
  br i64 1, label %"2", label %"3"

"1":                                              ; No predecessors!
  br label %"4"

"2":                                              ; preds = %"0"
  %a = alloca i64
  store i64 2, i64* %a
  br label %"3"

"3":                                              ; preds = %"2", %"0"

"4":                                              ; preds = %"1"

"5":                                              ; preds = %"6"

"6":                                              ; No predecessors!
  ret i64 3
  br label %"5"
}

